reference-implementation/lib/writable-stream.js

```diff
@@ -182,215 +190,242 @@ function WritableStreamAddWriteRequest(stream) {
-    readyPromiseIsPending = true;
-  }
+    WritableStreamRejectClosedPromiseIfAny(stream, stream._storedError);
 
-  if (wasAborted === true) {
-    stream._pendingAbortRequest._reject(reason);
-    stream._pendingAbortRequest = undefined;
+    return;
   }
 
-  if (state === 'errored') {
-    WritableStreamRejectClosedPromiseIfAny(stream);
+  assert(state === 'writable');
+
+  stream._state = 'errored';
+  stream._storedError = reason;
+
+  if (stream._pendingAbortRequest === undefined) {
+    const writer = stream._writer;
+    if (writer !== undefined) {
+      defaultWriterReadyPromiseResetToRejected(writer, reason);
+      writer._readyPromise.catch(() => {});
+    }
+
+    WritableStreamRejectClosedPromiseIfAny(stream, reason);
 
     return;
   }
 
-  assert(state === 'closing');
+  stream._pendingAbortRequest._reject(reason);
+  stream._pendingAbortRequest = undefined;
 
-  stream._state = 'errored';
-  stream._storedError = reason;
+  WritableStreamRejectClosedPromiseIfAny(stream, reason);
+}
+
+function WritableStreamCloseInQueueOrInFlight(stream) {
+  if (stream._closeRequest === undefined && stream._inFlightCloseRequest === undefined) {
+    return false;
+  }
 
-  if (wasAborted === false && stream._writer !== undefined) {
-    WritableStreamDefaultWriterEnsureReadyPromiseRejectedWith(stream._writer, reason, readyPromiseIsPending);
+  return true;
+}
+
+function WritableStreamHasOperationMarkedInFlight(stream) {
+  if (stream._inFlightWriteRequest === undefined && stream._inFlightCloseRequest === undefined) {
+    return false;
   }
 
-  WritableStreamRejectClosedPromiseIfAny(stream);
+  return true;
+}
+
+function WritableStreamMarkCloseRequestInFlight(stream) {
+  assert(stream._inFlightCloseRequest === undefined);
+  assert(stream._closeRequest !== undefined);
+  stream._inFlightCloseRequest = stream._closeRequest;
+  stream._closeRequest = undefined;
 }
 
-function WritableStreamMarkFirstWriteRequestPending(stream) {
-  assert(stream._pendingWriteRequest === undefined, 'there must be no pending write request');
+function WritableStreamMarkFirstWriteRequestInFlight(stream) {
+  assert(stream._inFlightWriteRequest === undefined, 'there must be no pending write request');
   assert(stream._writeRequests.length !== 0, 'writeRequests must not be empty');
-  stream._pendingWriteRequest = stream._writeRequests.shift();
+  stream._inFlightWriteRequest = stream._writeRequests.shift();
 }
 
-function WritableStreamRejectClosedPromiseIfAny(stream) {
+function WritableStreamRejectClosedPromiseIfAny(stream, reason) {
   const writer = stream._writer;
   if (writer !== undefined) {
-    defaultWriterClosedPromiseReject(writer, stream._storedError);
+    defaultWriterClosedPromiseReject(writer, reason);
     writer._closedPromise.catch(() => {});
   }
 }
 
-function WritableStreamRejectPromisesInReactionToError(stream) {
+function WritableStreamRejectPromisesInReactionToError(stream, reason) {
```

Comment:
As far as I can tell reason === storedError is always true. Why did you change it to be an argument?